nostd::list and nostd::pair

satac2 opened the pull request:
This is a simplified implementation of std::list and std::pair, I implemented the methods that I thought most essential. This should unblock the context api.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+  {
+    size_++;
+
+    T *temp = new T[size_];
```
We might want to store both a `size` and a `capacity`. In this way we could avoid the necessity of `new` and `delete` calls for each list manipulation (which is very costly).

satac2 replied:
Added a capacity member and resize function.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+{
+// Non-standard implementation of the list data structure.
+template <class T>
+class list
```
Naming this `vector` would  be more accurate.

satac2 replied:
Changed the name to vector.

pyohannes commented on api/test/nostd/BUILD:
```diff
@@ -74,3 +74,25 @@ cc_test(
     ],
 )
+
+cc_test(
```
Please also add this to the `CMakeList.txt` file.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+    return (memcmp(data_, other_list.data_, size_*sizeof(T)) == 0);
+  }
+
+  T &operator[](int index) { return data_[index]; }
```
Please use `size_t` instead of `int` (or even better, define a dedicated `size_type`).

satac2 replied:
Switched to size_t.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+      return false;
+    }
+    
+    return (memcmp(data_, other_list.data_, size_*sizeof(T)) == 0);
```
Rather use `std::equal(begin(), end(), other_list.begin())` here.

satac2 replied:
switched to std::equal.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+
+    data_ = new T[size_];
+
+    memcpy(data_, other_list.data_, other_list.size_ * sizeof(T));
```
This will not work when using smart pointers as list values.

satac2 replied:
Switched to std::copy.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+
+    data_ = new T[other_list.size_];
+
+    memcpy(data_, other_list.data_, other_list.size_ * sizeof(T));
```
This will not work when using smart pointers as list values.

satac2 replied:
Switched from memcpy to std::copy.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+  {
+    if (size_ == 0)
+    {
+      return data_[0];
```
That's not strictly necessary. In this case we'll have undefined behavior anyway.

satac2 replied:
Isn't it expected to fail when called on empty list?

lalitb replied:
yes you are correct. Please ignore.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+
+    data_ = new T[size_];
+
+    std::memcpy(data_, std::begin(init_list), init_list.size() * sizeof(T));
```
Might fail for smart pointers.

satac2 replied:
Switched to std::copy.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+#include <cstring>
+#include <iostream>
+#include <iterator>
+#include <memory>
```
Not all those header files are used.

satac2 replied:
Removed unnecessary headers.

pyohannes commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+
+    data_ = new T[size_];
+
+    std::memcpy(data_, &element, sizeof(T));
```
That's unsafe when used with smart pointers.

satac2 replied:
Switched to std::copy.

lalitb commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+    int previous_size = size_;
+    size_++;
+
+    T *temp = new T[size_];
```
Can we improve on this? re-allocating for every insert won't be efficient solution.

satac2 replied:
Added a resize function that will reduce the amount of allocations.

lalitb commented on api/include/opentelemetry/nostd/pair.h:
```diff
@@ -0,0 +1,52 @@
+  {
+    first_  = first;
+    second_ = second;
+  }
```
Please use initialization list.

satac2 replied:
Added initialization list.

lalitb commented on api/include/opentelemetry/nostd/pair.h:
```diff
@@ -0,0 +1,52 @@
+  {
+    first_  = T1();
+    second_ = T2();
+  }
```
Please use initialization list. Also, it will fail if T1 and T2 doesn't have default constructor ( which is acceptable I believe).

satac2 replied:
Removed T1 and T2 default constructors.

lalitb commented on api/include/opentelemetry/nostd/pair.h:
```diff
@@ -0,0 +1,52 @@
+
+  // Returns the second element.
+  T2 &second() { return second_; }
+
```
We can add make_pair function too to make this logically complete.

satac2 replied:
Added make_pair function to pair.h

maxgolov commented:
I might be misunderstanding the purpose of nostd::list and nostd::pair being necessary for Context API..

Can you consider using the existing types:
- **nostd::variant** or **opentelemetry::common::AttributeValue** - to express any value / attribute
- or std::map that'd be transformed to KeyValueIterableView<T> ( see here: https://github.com/open-telemetry/opentelemetry-cpp/blob/ed0a36e2906ab7eb03c3b0a4c8dea02eb78849ce/api/include/opentelemetry/trace/span.h#L109 )
- **nostd::span** - to express a collection of values

We had a related discussion in Gitter: https://gitter.im/open-telemetry/opentelemetry-cpp

Adding more classes to nostd namespace comes at a maintenance cost. I'd suggest we don't do this, and instead reuse the existing implementations whenever possible.

satac2 commented:
> I might be misunderstanding the purpose of nostd::list and nostd::pair being necessary for Context API..
> 
> Can you consider using the existing types:
> 
> * **nostd::variant** or **opentelemetry::common::AttributeValue** - to express any value / attribute
> * or std::map that'd be transformed to KeyValueIterableView ( see here: https://github.com/open-telemetry/opentelemetry-cpp/blob/ed0a36e2906ab7eb03c3b0a4c8dea02eb78849ce/api/include/opentelemetry/trace/span.h#L109
>    )
> * **nostd::span** - to express a collection of values
> 
> We had a related discussion in Gitter: https://gitter.im/open-telemetry/opentelemetry-cpp
> 
> Adding more classes to nostd namespace comes at a maintenance cost. I'd suggest we don't do this, and instead reuse the existing implementations whenever possible.

satac2 commented:
> I might be misunderstanding the purpose of nostd::list and nostd::pair being necessary for Context API..
> 
> Can you consider using the existing types:
> 
> * **nostd::variant** or **opentelemetry::common::AttributeValue** - to express any value / attribute
> * or std::map that'd be transformed to KeyValueIterableView ( see here: https://github.com/open-telemetry/opentelemetry-cpp/blob/ed0a36e2906ab7eb03c3b0a4c8dea02eb78849ce/api/include/opentelemetry/trace/span.h#L109
>    )
> * **nostd::span** - to express a collection of values
> 
> We had a related discussion in Gitter: https://gitter.im/open-telemetry/opentelemetry-cpp
> 
> Adding more classes to nostd namespace comes at a maintenance cost. I'd suggest we don't do this, and instead reuse the existing implementations whenever possible.

I can try using nostd::span to do it. Can we use std::map?

satac2 commented:
> I might be misunderstanding the purpose of nostd::list and nostd::pair being necessary for Context API..
> 
> Can you consider using the existing types:
> 
> * **nostd::variant** or **opentelemetry::common::AttributeValue** - to express any value / attribute
> * or std::map that'd be transformed to KeyValueIterableView ( see here: https://github.com/open-telemetry/opentelemetry-cpp/blob/ed0a36e2906ab7eb03c3b0a4c8dea02eb78849ce/api/include/opentelemetry/trace/span.h#L109
>    )
> * **nostd::span** - to express a collection of values
> 
> We had a related discussion in Gitter: https://gitter.im/open-telemetry/opentelemetry-cpp
> 
> Adding more classes to nostd namespace comes at a maintenance cost. I'd suggest we don't do this, and instead reuse the existing implementations whenever possible.

I'm not sure those attributeValue, variant and span, provide me with what I am looking for in order to track a list of keys and values. Also, my understanding was that std::map was off the table, if I could use that it would be ideal. However, I figure that the next best thing is a list of pairs<std::string, AttributeValue>. Can I use std::map? @maxgolov

pyohannes commented:
> Would you consider using existing types (nostd::span, nostd::variant, KeyValueIterableView, AttributeValue).

But how would we manage the lifetime of a `nostd::span` view on a vector? Who would own the vector an be responsible for freeing the memory?

pyohannes commented:
> Can I use std::map?

All the header-only API stuff might be shared across DSOs compiled with different STL libraries. I think in this scenario `std::map` as a class member could ABI problems.

reyang commented:
> > Can I use std::map?

Not in the public APIs. #134

satac2 commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+    // Copy the original data that comes before the insertion.
+    std::memcpy(temp, data_, insert_index * sizeof(T));
+    // Copy the inserted data.
+    std::memcpy(temp + insert_index, &element, sizeof(T));
```
Removed mempcys from everywhere except the resize function.

satac2 replied:
I added a test for smart pointers using std::shared_ptr, but the cmake tests rejected my use of shared_ptr so I removed them and I don't think it makes sense to add a unique_ptr to a vector, so I do not currently have tests for smart pointers.

satac2 commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+  list()
+  {
+    size_ = 0;
+    data_ = nullptr;
```
Added when possible.

satac2 commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+#include <iterator>
+#include <memory>
+
+#include "opentelemetry/nostd/utility.h"
```
I had yet to implement a const iterator, so it was unusable, however I have and am now using it.

satac2 commented on api/include/opentelemetry/nostd/list.h:
```diff
@@ -0,0 +1,254 @@
+  }
+
+  // Returns the element at the front of the list.
+  T &front() { return data_[0]; }
```
Isn't it expected to fail when called on an empty list?

lalitb replied:
Yes correct. Please ignore my comment here.

satac2 commented:
> I added some comments inline.
> 
> I think we should strip this down to the absolute minimum necessary. I'm not sure whether we'll need `insert`, `push_front` and `pop_front` for our purposes. If there's no initial requirement for this for the `Context` implementation, I'd start without them and add them when needed. That's just my opinion though, the idea is mostly to eliminate possible bugs and testing needs.

I removed pop_front and push_front, but left in insert.

maxgolov commented:
@satac2 @reyang @pyohannes 

You can use **std::map** in the client app, just not on API surface.. which will be transformed to key-value iterable.

Example:
https://github.com/maxgolov/opentelemetry-cpp/blob/131b2e6474347b3f14d9e976f0418eb1e163e789/examples/TraceStreamer/main.cpp#L42

https://github.com/maxgolov/opentelemetry-cpp/blob/131b2e6474347b3f14d9e976f0418eb1e163e789/examples/TraceStreamer/main.cpp#L102

After I merge the change for comparator, you can also use **nostd::string_view** as key --  as in std::map<**nostd::string_view**, common::AttributeValue>.

Same concept can be applied to express a **nostd::span** of **common::AttributeValue** , as in contiguous container of variants objects. The only attribute of **std::list** is ability to contain non-contiguous objects. What is the usage example / benefits when we require this on API surface?

maxgolov commented:
> But how would we manage the lifetime of a `nostd::span` view on a vector? Who would own the vector and be responsible for freeing the memory?

@pyohannes - I don't understand the problem or concern here. SDK does not own the memory: similar to what you're doing with SpanData, you'd either copy or serialize into another alternate format right away on caller's thread. Since you'd be needing a memcpy for some scenarios, same comment applies - in some implementations of SDK API surface, the SDK may need to copy, on others - don't. The app that calls into API owns the memory, and guarantees that the object is there until API call returns.

Then under the hood - SDK either:
- right away streams the parameters to zero-copy sync exporter (e.g. serializing on calling thread), or
- async exporter would create a copy of all attributes and events passed to it (for subsequent batching).
